### PR TITLE
journal: convert --list-boots to a table

### PR DESCRIPTION
**Pull request by mrc0mmand**
so it can be output as JSON as well.

```
$ build-san/journalctl --list-boots --file boot-test.journal
#  BOOT ID                          FIRST ENTRY                 LAST ENTRY
-3 39d66eb1925f4d01b8464d502650a714 Sat 2022-03-05 15:20:33 CET Sat 2022-03-05 16:19:21 CET
-2 5dffeb08a27344d5ae9e2fc244bbcbc5 Fri 2022-03-11 17:23:57 CET Fri 2022-03-11 17:38:31 CET
-1 c8ebd52915b642c39eda4bf00f864f79 Fri 2022-03-11 17:38:41 CET Fri 2022-03-11 20:03:46 CET
 0 00bcba97c7094fa88cc5d1cf2a389057 Sat 2022-03-12 20:39:08 CET Sat 2022-03-12 20:40:18 CET

$ build-san/journalctl --list-boots --file boot-test.journal -q
-3 39d66eb1925f4d01b8464d502650a714 Sat 2022-03-05 15:20:33 CET Sat 2022-03-05 16:19:21 CET
-2 5dffeb08a27344d5ae9e2fc244bbcbc5 Fri 2022-03-11 17:23:57 CET Fri 2022-03-11 17:38:31 CET
-1 c8ebd52915b642c39eda4bf00f864f79 Fri 2022-03-11 17:38:41 CET Fri 2022-03-11 20:03:46 CET
 0 00bcba97c7094fa88cc5d1cf2a389057 Sat 2022-03-12 20:39:08 CET Sat 2022-03-12 20:40:18 CET

$ build-san/journalctl --list-boots --file boot-test.journal -o json-pretty
[
        {
                "#" : "-3",
                "boot_id" : "39d66eb1925f4d01b8464d502650a714",
                "first_entry" : 1646490033438495,
                "last_entry" : 1646493561047353
        },
        {
                "#" : "-2",
                "boot_id" : "5dffeb08a27344d5ae9e2fc244bbcbc5",
                "first_entry" : 1647015837289036,
                "last_entry" : 1647016711595489
        },
        {
                "#" : "-1",
                "boot_id" : "c8ebd52915b642c39eda4bf00f864f79",
                "first_entry" : 1647016721056382,
                "last_entry" : 1647025426397414
        },
        {
                "#" : "0",
                "boot_id" : "00bcba97c7094fa88cc5d1cf2a389057",
                "first_entry" : 1647113948506002,
                "last_entry" : 1647114018943637
        }
]
```

Resolves: #14625

### CODE REVIEW

File: src/journal/journalctl.c [hunk 1]
```diff
@@ -1452,23 +1458,36 @@ static int list_boots(sd_journal *j) {
         if (count == 0)
                 return count;
 
-        pager_open(arg_pager_flags);
+        table = table_new("#", "boot id", "first entry", "last entry");
```
Comment (yuwata) [inline src/journal/journalctl.c]: `#` is not a good label especially for json. How about to use `"index"` or `"boot index"` at least when printed in json format?

**Review by yuwata — APPROVED** [review]
Looks nice, but one comment.

File: src/journal/journalctl.c (hunk 1 above)
Comment (mrc0mmand) [reply src/journal/journalctl.c]: Would "idx" work when used in both cases?

```
$ build-san/journalctl --list-boots --file boot-test.journal 
IDX BOOT ID                          FIRST ENTRY                 LAST ENTRY                 
-3  39d66eb1925f4d01b8464d502650a714 Sat 2022-03-05 15:20:33 CET Sat 2022-03-05 16:19:21 CET
-2  5dffeb08a27344d5ae9e2fc244bbcbc5 Fri 2022-03-11 17:23:57 CET Fri 2022-03-11 17:38:31 CET
-1  c8ebd52915b642c39eda4bf00f864f79 Fri 2022-03-11 17:38:41 CET Fri 2022-03-11 20:03:46 CET
 0  00bcba97c7094fa88cc5d1cf2a389057 Sat 2022-03-12 20:39:08 CET Sat 2022-03-12 20:40:18 CET

$ build-san/journalctl --list-boots --file boot-test.journal -o json-pretty
[
        {
                "idx" : "-3",
                "boot_id" : "39d66eb1925f4d01b8464d502650a714",
                "first_entry" : 1646490033438495,
                "last_entry" : 1646493561047353
        },
        {
                "idx" : "-2",
                "boot_id" : "5dffeb08a27344d5ae9e2fc244bbcbc5",
                "first_entry" : 1647015837289036,
                "last_entry" : 1647016711595489
        },
        {
                "idx" : "-1",
                "boot_id" : "c8ebd52915b642c39eda4bf00f864f79",
                "first_entry" : 1647016721056382,
                "last_entry" : 1647025426397414
        },
        {
                "idx" : "0",
                "boot_id" : "00bcba97c7094fa88cc5d1cf2a389057",
                "first_entry" : 1647113948506002,
                "last_entry" : 1647114018943637
        }
]
```

I mean, there's an "extra" space in the "IDX" column (since all indexes in my test journal are two-chars wide at most), but it uses the same _label_ in both cases.

Comment (yuwata) [reply src/journal/journalctl.c]: Still I am not a fan of abbreviated name in json format. `idx` or `#` is OK for table format. How about the following?
```
table = table_new(OUTPUT_MODE_IS_JSON(arg_output) ? "index" : "idx", "boot id", "first entry", "last entry");
```

File: src/journal/journalctl.c [hunk 2]
```diff
@@ -1452,23 +1458,36 @@ static int list_boots(sd_journal *j) {
...
-
-                printf("% *i " SD_ID128_FORMAT_STR " %s—%s\n",
-                       w, i - count + 1,
-                       SD_ID128_FORMAT_VAL(id->id),
-                       format_timestamp_maybe_utc(a, sizeof(a), id->first),
-                       format_timestamp_maybe_utc(b, sizeof(b), id->last));
+                char idx[w + 1];
+
+                /* Right-align the index unless JSON is requested */
+                xsprintf(idx, "%*i", OUTPUT_MODE_IS_JSON(arg_output) ? 0 : w, i - count + 1);
```
Comment (yuwata) [inline src/journal/journalctl.c]: Please use TABLE_INT for `i`, and use `TABLE_SET_ALIGN_PERCENT` to set the alignment. Then, the alignment is gracefully ignored when the output mode is json.

Comment (mrc0mmand) [reply src/journal/journalctl.c]: Ah, sweet, that's exactly something I was looking for when I was dealing with the alignment, but couldn't find anything relevant.

**Comment by mrc0mmand** [comment]
While trying to use `TABLE_INT`, I started hitting heap-buffer-overflows, which were caused by `DECIMAL_STR_WIDTH()` not account for the minus sign in negative numbers, thus returning a too-short buffer for the resulting number (if negative, like is the case with previous boot indexes).

I'm not sure if the solution I proposed in the second commit (or first commit in this case) is correct, but I couldn't do a simple `if (x < 0)` check, since that triggers gcc warnings when doing this comparison on a unsigned type.

@yuwata thanks for all the suggestions and PTAL!

File: src/basic/macro.h [hunk 3]
```diff
@@ -323,6 +323,8 @@ static inline int __coverity_check_and_return__(int condition) {
         ({                                              \
                 typeof(x) _x_ = (x);                    \
                 size_t ans = 1;                         \
+                if (IS_SIGNED_INTEGER_TYPE(_x_))        \
+                        ans++;                          \
```
Comment (yuwata) [inline src/basic/macro.h]: Ouch!

File: src/journal/journalctl.c [hunk 4]
```diff
@@ -1441,8 +1445,10 @@ static int get_boots(
 }
 
 static int list_boots(sd_journal *j) {
-        int w, i, count;
+        _cleanup_(table_unrefp) Table *table = NULL;
+        int i, count;
         BootId *id, *all_ids;
+        int r;
```
Comment (yuwata) [inline src/journal/journalctl.c]: The declarations of `int` can be merged.

File: src/journal/journalctl.c [hunk 5]
```diff
@@ -33,6 +33,7 @@
 #include "dissect-image.h"
```
Comment (yuwata) [inline src/journal/journalctl.c:33]: The commit message includes unwanted comments.

File: src/basic/macro.h (hunk 3 above)
Comment (yuwata) [inline src/basic/macro.h]: Not a big matter, but is it possible to do like the following?
```
                if (IS_SIGNED_INTEGER_TYPE(_x_))        \
                        ans += _x_ < 0;
```

File: src/journal/journalctl.c (hunk 5 above)
Comment (mrc0mmand) [reply src/journal/journalctl.c:33]: Whoops, thanks!

File: src/basic/macro.h (hunk 3 above)
Comment (mrc0mmand) [reply src/basic/macro.h]: Unfortunately, that still makes gcc complain:

```
[443/1827] Compiling C object src/shared/libsystemd-shared-250.a.p/json.c.o
In file included from ../src/basic/alloc-util.h:9,
                 from ../src/shared/json.c:12:
../src/shared/json.c: In function ‘print_source’:
../src/basic/macro.h:327:36: warning: comparison of unsigned expression in ‘< 0’ is always false [-Wtype-limits]
  327 |                         ans += _x_ < 0;                 \
      |                                    ^
../src/shared/json.c:1466:17: note: in expansion of macro ‘DECIMAL_STR_WIDTH’
 1466 |                 DECIMAL_STR_WIDTH(v->source->max_line) :
      |  
```

Comment (mrc0mmand) [reply src/basic/macro.h]: I also tried to use `DISABLE_WARNING_TYPE_LIMITS` and `REENABLE_WARNING` around the condition, but that just triggers a different compilation issue.

Comment (yuwata) [reply src/basic/macro.h]: OK. Let's keep this as is. One extra buffer space should not cause anything serious.

**Review by yuwata — APPROVED** [review]
LGTM.

File: src/basic/macro.h (hunk 3 above)
Comment (medhefgo) [reply src/basic/macro.h]: `ans += __builtin_choose_expr(IS_SIGNED_INTEGER_TYPE(_x_), 1, 0)`; should do the trick

Comment (mrc0mmand) [reply src/basic/macro.h]: Indeed, thanks for the hint!

File: src/basic/macro.h [hunk 6]
```diff
@@ -319,13 +319,14 @@ static inline int __coverity_check_and_return__(int condition) {
...
-                typeof(x) _x_ = (x);                    \
-                size_t ans = 1;                         \
-                while ((_x_ /= 10) != 0)                \
-                        ans++;                          \
-                ans;                                    \
+#define DECIMAL_STR_WIDTH(x)                                                     \
+        ({                                                                       \
+                typeof(x) _x_ = (x);                                             \
+                size_t ans = 1;                                                  \
+                ans += __builtin_choose_expr(IS_SIGNED_INTEGER_TYPE(_x_), 1, 0); \
```
Comment (medhefgo) [inline src/basic/macro.h]: Might as well combine these two lines (have the choose_expr return 2 or 1)...

File: src/basic/macro.h (hunk 3 above)
Comment (yuwata) [reply src/basic/macro.h]: > `ans += __builtin_choose_expr(IS_SIGNED_INTEGER_TYPE(_x_), 1, 0)`; should do the trick

Hm? What is the merit of using `__builtin_choose_expr()`? Why not to use the following?
```
size_t ans = 1 + IS_SIGNED_INTEGER_TYPE(_x_);
```

Comment (medhefgo) [reply src/basic/macro.h]: You have a point. Though, `IS_SIGNED_INTEGER_TYPE(_x_) ? 2 : 1` is preferred by coding style.

**Comment by mrc0mmand** [comment]
Updated, thank you both for the suggestions!

**Comment by mrc0mmand** [comment]
Also, extended the format-table tests, which now correctly trigger the heap-buffer-overflow when [e3dd9ea](https://github.com/systemd/systemd/pull/22721/commits/e3dd9ea8ea4510221f73071ad30ee657ca77565d) is reverted:

```
/* test_table */                                                                                                       
=================================================================                                                      
==3590066==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x602000005812 at pc 0x7f4a07a4cb2e bp 0x7ffc1e92e930 sp 0x7ffc1e92e0e0                                                                                                                                                                                                                                                    
WRITE of size 3 at 0x602000005812 thread T0                
    #0 0x7f4a07a4cb2d in vsprintf (/lib64/libasan.so.6+0x5cb2d)                                                        
    #1 0x7f4a07a4cd1e in __interceptor_sprintf (/lib64/libasan.so.6+0x5cd1e)                                                                                                                   
    #2 0x7f4a0662642e in table_data_format ../src/shared/format-table.c:1529                                           
    #3 0x7f4a0662ace5 in table_data_requested_width_height ../src/shared/format-table.c:1854                           
    #4 0x7f4a0662d924 in table_print ../src/shared/format-table.c:2053                                                 
    #5 0x7f4a06632ac1 in table_format ../src/shared/format-table.c:2393                                                
    #6 0x4080dd in test_table ../src/test/test-format-table.c:416                                                      
    #7 0x402e20 in run_test_table ../src/shared/tests.h:104                                                            
    #8 0x4090c0 in main ../src/test/test-format-table.c:543                                                            
    #9 0x7f4a04fb155f in __libc_start_call_main (/lib64/libc.so.6+0x2d55f)                                             
    #10 0x7f4a04fb160b in __libc_start_main_alias_1 (/lib64/libc.so.6+0x2d60b)                                         
    #11 0x402314 in _start (/home/mrc0mmand/repos/systemd/build-san/test-format-table+0x402314)                        
                                                                                                                       
0x602000005812 is located 0 bytes to the right of 2-byte region [0x602000005810,0x602000005812)                        
allocated by thread T0 here:          
    #0 0x7f4a07a9e91f in __interceptor_malloc (/lib64/libasan.so.6+0xae91f)                                                             
    #1 0x7f4a06613760 in malloc_multiply ../src/basic/alloc-util.h:104                                                                  
    #2 0x7f4a066262e9 in table_data_format ../src/shared/format-table.c:1525                                                            
    #3 0x7f4a0662ace5 in table_data_requested_width_height ../src/shared/format-table.c:1854                                            
    #4 0x7f4a0662d924 in table_print ../src/shared/format-table.c:2053                                                                  
    #5 0x7f4a06632ac1 in table_format ../src/shared/format-table.c:2393                                                                 
    #6 0x4080dd in test_table ../src/test/test-format-table.c:416                                                                       
    #7 0x402e20 in run_test_table ../src/shared/tests.h:104                                                                             
    #8 0x4090c0 in main ../src/test/test-format-table.c:543                                                                             
    #9 0x7f4a04fb155f in __libc_start_call_main (/lib64/libc.so.6+0x2d55f)                                                              
                                                                                                                                        
SUMMARY: AddressSanitizer: heap-buffer-overflow (/lib64/libasan.so.6+0x5cb2d) in vsprintf                                               
Shadow bytes around the buggy address:
  0x0c047fff8ab0: fa fa fd fa fa fa fd fa fa fa fd fd fa fa fd fa                                                                                                                              
  0x0c047fff8ac0: fa fa fd fd fa fa fd fd fa fa fd fa fa fa fd fd                                                                                                                              
  0x0c047fff8ad0: fa fa fd fd fa fa fd fa fa fa fd fa fa fa fd fa                                                      
  0x0c047fff8ae0: fa fa fd fa fa fa fd fa fa fa fd fa fa fa fd fd                                                                                                                                                                              
  0x0c047fff8af0: fa fa 04 fa fa fa 04 fa fa fa 06 fa fa fa 05 fa                                                                                                                              
=>0x0c047fff8b00: fa fa[02]fa fa fa fa fa fa fa fa fa fa fa fa fa                                                      
  0x0c047fff8b10: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa                                                      
  0x0c047fff8b20: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa                                                      
  0x0c047fff8b30: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa                                                                                                                                                                              
  0x0c047fff8b40: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa                                                                                                                              
  0x0c047fff8b50: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa                                                                                                                              
Shadow byte legend (one shadow byte represents 8 application bytes):                                                                                                                           
  Addressable:           00                                         
  Partially addressable: 01 02 03 04 05 06 07                       
  Heap left redzone:       fa                                       
  Freed heap region:       fd                                       
  Stack left redzone:      f1                                       
  Stack mid redzone:       f2                                       
  Stack right redzone:     f3                                       
  Stack after return:      f5                                       
  Stack use after scope:   f8                                       
  Global redzone:          f9                                                                                                                                                                                                                                                    
  Global init order:       f6                                                                  
  Poisoned by user:        f7                                                                  
  Container overflow:      fc                                                                  
  Array cookie:            ac                                                                  
  Intra object redzone:    bb                                                                  
  ASan internal:           fe                                                                  
  Left alloca redzone:     ca                                                                  
  Right alloca redzone:    cb                                                                  
  Shadow gap:              cc                                                                  
==3590066==ABORTING                                                                            
Aborted (core dumped)                                                                        
```

**Review by medhefgo — APPROVED** [review]

**Comment by yuwata** [comment]
The first commit needs to be backported to stable repo.

File: src/journal/journalctl.c [hunk 7]
```diff
@@ -1452,23 +1457,30 @@ static int list_boots(sd_journal *j) {
         if (count == 0)
                 return count;
 
-        pager_open(arg_pager_flags);
+        table = table_new(OUTPUT_MODE_IS_JSON(arg_output) ? "index" : "idx", "boot id", "first entry", "last entry");
```
Comment (poettering) [inline src/journal/journalctl.c:1460]: please use table_set_json_field_name() to override the JSON field name for the first column, instead of the ternary op

i.e. we should try to generate always the same internal table representation if we can, and if output needs to be tweaked move that to a secondary step. table_set_json_field_name() allows doing that by overriding for specific columns what the actual field name shall be in case the column title is not suitable.

File: src/journal/journalctl.c [hunk 8]
```diff
@@ -577,6 +579,8 @@ static int parse_argv(int argc, char *argv[]) {
                         if (IN_SET(arg_output, OUTPUT_EXPORT, OUTPUT_JSON, OUTPUT_JSON_PRETTY, OUTPUT_JSON_SSE, OUTPUT_JSON_SEQ, OUTPUT_CAT))
                                 arg_quiet = true;
 
+                        if (OUTPUT_MODE_IS_JSON(arg_output))
+                                arg_json_format_flags = output_mode_to_json_format_flags(arg_output) | JSON_FORMAT_COLOR_AUTO;
```
Comment (poettering) [inline src/journal/journalctl.c:583]: i think this should have an else branch that resets arg_json_format_flags back to JSON_FORMAT_OFF.

`journalctl -o json-pretty -o export` might otherwise produce unexpected output (i.e. I'd expect the second -o overrides the first in full, but without the else it would not)

**Comment by poettering** [comment]
mostly lgtm, but two more nits I foud in post-merge review
